drivers/iio/frequency/adf41513.c

```diff
@@ -0,0 +1,1196 @@
+		if (ret < 0) {
+			dev_err(&st->spi->dev, "Failed to set power-up frequency: %d\n", ret);
+			return ret;
+		}
+	} else {
+		/* init powered down */
+		st->regs[ADF41513_REG6] |= ADF41513_REG6_POWER_DOWN(1);
+		st->regs[ADF41513_REG12] &= ~ADF41513_REG12_LE_SELECT_MSK;
+
+		/* write initial configuration */
+		ret = adf41513_sync_config(st, ADF41513_SYNC_ALL);
+		if (ret < 0) {
+			dev_err(&st->spi->dev, "Failed to write initial config: %d\n", ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static void adf41513_power_down(void *data)
+{
+	struct adf41513_state *st = data;
+
+	guard(mutex)(&st->lock);
+	adf41513_suspend(st);
+
+	if (st->chip_enable)
+		gpiod_set_value_cansleep(st->chip_enable, 0);
+}
+
+static int adf41513_pm_suspend(struct device *dev)
+{
+	struct iio_dev *indio_dev = dev_get_drvdata(dev);
+	struct adf41513_state *st = iio_priv(indio_dev);
+
+	guard(mutex)(&st->lock);
+	return adf41513_suspend(st);
+}
+
+static int adf41513_pm_resume(struct device *dev)
+{
+	struct iio_dev *indio_dev = dev_get_drvdata(dev);
+	struct adf41513_state *st = iio_priv(indio_dev);
+
+	guard(mutex)(&st->lock);
+	return adf41513_resume(st);
+}
+
+static int adf41513_probe(struct spi_device *spi)
+{
+	struct adf41513_platform_data *pdata;
+	struct iio_dev *indio_dev;
+	struct adf41513_state *st;
+	struct clk *ref_clk;
+
+	int ret;
+
+	/* Parse device tree or platform data */
+	if (dev_fwnode(&spi->dev)) {
+		pdata = adf41513_parse_dt(&spi->dev);
+		if (IS_ERR(pdata))
+			return PTR_ERR(pdata);
+	} else {
+		pdata = spi->dev.platform_data;
+		if (!pdata) {
+			dev_err(&spi->dev, "No platform data provided\n");
+			return -EINVAL;
+		}
+	}
+
+	/* Get reference clock */
+	if (!pdata->clkin) {
+		ref_clk = devm_clk_get_prepared(&spi->dev, "clkin");
+		if (IS_ERR(ref_clk))
+			return -EPROBE_DEFER;
+	}
+
+	indio_dev = devm_iio_device_alloc(&spi->dev, sizeof(*st));
+	if (!indio_dev)
+		return -ENOMEM;
+
+	st = iio_priv(indio_dev);
+	st->spi = spi;
+	st->pdata = pdata;
+	st->ref_clk = ref_clk;
+	st->freq_resolution_uhz = pdata->freq_resolution_uhz;
+	st->device_id = (u32)spi_get_device_id(spi)->driver_data;
+
+	spi_set_drvdata(spi, indio_dev);
+
+	/* vcc regulator */
+	ret = devm_regulator_get_enable(&spi->dev, "vcc");
+	if (ret) {
+		return dev_err_probe(&spi->dev, ret,
+				     "failed to get and enable vcc regulator\n");
+	}
+
+	/* get chip enable gpio */
+	st->chip_enable = devm_gpiod_get_optional(&spi->dev, "chip-enable", GPIOD_OUT_HIGH);
+	if (IS_ERR(st->chip_enable)) {
+		dev_err(&spi->dev, "fail to request chip enable GPIO\n");
+		return PTR_ERR(st->chip_enable);
+	}
+
+	/* get lock detect gpio */
+	st->lock_detect = devm_gpiod_get_optional(&spi->dev, "lock-detect", GPIOD_IN);
+	if (IS_ERR(st->lock_detect)) {
+		dev_err(&spi->dev, "fail to request lock detect GPIO\n");
+		return PTR_ERR(st->lock_detect);
+	}
+
+	/* set reference frequency */
+	if (st->ref_clk)
+		st->ref_freq_hz = clk_get_rate(st->ref_clk);
+	else
+		st->ref_freq_hz = pdata->clkin;
```

Comment:
```suggestion
	if (pdata->clkin)
		st->ref_freq_hz = pdata->clkin;
	else
		st->ref_freq_hz = clk_get_rate(st->ref_clk);
```

flows a little more naturally IMO, since ref_clk is set if clkin is not set